### ISSUE TITLE
Issue codes for explicit failures

### ISSUE THREAD

**Issue by rolve** [LorenzoBettini/xsemantics#58]
I find the issue codes that are generated for the type system rules very useful for testing. However, I find myself using explicit failures more and more often, because they allow me to provide detailed error information. Unfortunately, explicit failures contain no issue codes!

It would be nice if these failures contained the issue code that corresponds to the containing rule. Even better, if the programmer could provide a more specific issue code manually.

**Comment by LorenzoBettini** [comment]
Yes, that's an interesting feature; I'll work on that as soon as possible.

It is clear what you mean by specifying issuecode on explicit failure, but could you please provide further information concerning the part "the programmer could provide a more specific issue code manually"?

thanks in advance

**Comment by rolve** [comment]
Thank you for your answer. I hope I can make this clearer. My idea is that, by default, explicit failures contain the containing rule's issue code, but the programmer can provide a different one explicitly, like so:

    fail
        error "type mismatch: bla bla bla"
        code TYPE_MISMATCH // this could be any string
        source s
